### PR TITLE
Persistent match record implementation

### LINKED ISSUE #1546: Create Persistent Match Records
When a match is found, create a record that can be used to reference the match again later.  For this story, we can create a new record every time a query results in a match, but that behavior will be refined in [story link]

The match record should include:
- [ ] A unique match ID (should follow the same formatting guidance as we used for lookup IDs)
- [ ] The hash representing the deidentified PII (name, DOB, SSN) - this will be used to determine whether or not to create a new match.
- [ ] The date + time when the match was found
- [ ] The states involved in the match. (At this stage, we can assume only two states per match, but this will likely change eventually)
- [ ] If the match is invalid (this should default to false)
- [ ] A status for the match.  This should default to "Open".

### PR DESCRIPTION
**Pull request by kingcomma**
- Service for adding records
- DAO for accessing database
- Builder for creating a `MatchRecordDbo` from match event data
- Service for generating match IDs

Partially addresses #1546. Final work will be implemented once this PR is reconciled with #1737. Rough plan at this point is for the orchestrator to loop through all matches and call `MatchRecordService.AddRecord()` for each.

Also, this work only addresses the need to create records for now. Retrieving/reconciling records will be addressed with #1549.

### CODE REVIEW

File: match/src/Piipan.Match/Piipan.Match.Core/Services/MatchRecordService.cs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+            // Match IDs are generated randomly from a large pool, but collision
+            // is still possible. Retry failed inserts a limited number of times.
+            var retries = 0;
+            while (retries < InsertRetries)
+            {
+                try
+                {
+                    return await _matchRecordDao.AddRecord(matchRecordDbo);
+                }
+                catch (PostgresException ex)
```
Comment (timoballard) [inline match/src/Piipan.Match/Piipan.Match.Core/Services/MatchRecordService.cs:52]: _nitpick_:  can we have the DAO catch the `PostgresException` and rethrow it either as something more generic (e.g. `InvalidOperationException`) or as something custom (e.g. `DuplicateMatchRecordException`)? That way the DAO isn't leaking implementation details (in this case, that we're using PSQL) into the service layer

File: match/src/Piipan.Match/Piipan.Match.Core/Models/ActiveMatchRecordBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,48 @@
+using System.Text.Json;
+using Piipan.Match.Orchestrator;
+
+namespace Piipan.Match.Core.Models
```
Comment (timoballard) [inline match/src/Piipan.Match/Piipan.Match.Core/Models/ActiveMatchRecordBuilder.cs:4]: _suggestion (non-blocking)_: I think a separate `Builders` namespace/directory would make sense here

File: match/src/Piipan.Match/Piipan.Match.Core/Models/IActiveMatchRecordBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+using Piipan.Match.Orchestrator;
+
+namespace Piipan.Match.Core.Models
+{
+    public interface IActiveMatchRecordBuilder
+    {
+        IActiveMatchRecordBuilder SetMatch(RequestPerson input, ParticipantRecord match);
+        IActiveMatchRecordBuilder SetStates(string stateA, string stateB);
+
+        MatchRecordDbo GetRecord();
```
Comment (timoballard) [inline match/src/Piipan.Match/Piipan.Match.Core/Models/IActiveMatchRecordBuilder.cs:11]: _question_: can this return an `IMatchRecord`?

File: match/tests/Piipan.Match.Core.Tests/Models/MatchRecordDboTests.cs [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+            };
+
+
+            // Act / Assert
+            Assert.True(record.Equals(recordMatch));
+            Assert.Equal(record.GetHashCode(), recordMatch.GetHashCode());
+        }
+
+        [Fact]
+        public void Equals_Match()
```
Comment (timoballard) [inline match/tests/Piipan.Match.Core.Tests/Models/MatchRecordDboTests.cs:336]: _nitpick_: I think if this test also showed that equality doesn't depend on the `createdAt` property, the test coverage for this PR might get to 100% (from 99 😁)

**Review by timoballard — COMMENTED** [review]
Took a quick first pass - looks great! Left a few comments

**Comment by kingcomma** [comment]
@timoballard Really appreciate the review and suggestions — I think they made for some solid improvements! I pushed a commit that should address everything, and I also added in some documentation on the public classes/methods that I had intended to include before.

**Review by timoballard — APPROVED** [review]
:amaze:

**Pull request merged**